pennylane/fourier/advanced_spectrum.py

```diff
@@ -0,0 +1,365 @@
+        argnum (list[int]): Numerical indices for arguments with respect to which
+            to compute the spectrum
+        decimals (int): number of decimals to which to round frequencies.
+        num_pos (int): Number of additional random positions at which to evaluate the
+            Jacobian of the preprocessing and test that it is constant.
+            Setting ``num_pos=0`` will deactivate the test.
+
+    Returns:
+        function: Function which accepts the same arguments as the QNode.
+            When called, this function will return a dictionary of dictionaries
+            containing the frequency spectra per QNode parameter.
+
+    **Details**
+
+    A circuit that returns an expectation value of a Hermitian observable which depends on
+    :math:`N` scalar inputs :math:`x_j` can be interpreted as a function
+    :math:`f: \mathbb{R}^N \rightarrow \mathbb{R}` (as the observable is Hermitian,
+    the expectation value is real-valued).
+    This function can always be expressed by a Fourier-type sum
+
+    .. math::
+
+        \sum \limits_{\omega_1\in \Omega_1} \dots \sum \limits_{\omega_N \in \Omega_N}
+        c_{\omega_1,\dots, \omega_N} e^{-i x_1 \omega_1} \dots e^{-i x_N \omega_N}
+
+    over the *frequency spectra* :math:`\Omega_j \subseteq \mathbb{R},`
+    :math:`j=1,\dots,N`. Each spectrum has the property that
+    :math:`0 \in \Omega_j`, and the spectrum is symmetric
+    (i.e., for every :math:`\omega \in \Omega_j` we have that :math:`-\omega \in\Omega_j`).
+    If all frequencies are integer-valued, the Fourier sum becomes a *Fourier series*.
+
+    As shown in `Vidal and Theis (2019) <https://arxiv.org/abs/1901.11434>`_ and
+    `Schuld, Sweke and Meyer (2020) <https://arxiv.org/abs/2008.08605>`_,
+    if an input :math:`x_j, j = 1 \dots N`,
+    only enters into single-parameter gates of the form :math:`e^{-i x_j G}`
+    (where :math:`G` is a Hermitian generator),
+    the frequency spectrum :math:`\Omega_j` is fully determined by the eigenvalues
+    of the generators :math:`G`. In many situations, the spectra are limited
+    to a few frequencies only, which in turn limits the function class that the circuit
+    can express.
+
+    The ``advanced_spectrum`` function computes all frequencies that will
+    potentially appear in the sets :math:`\Omega_1` to :math:`\Omega_N`.
+
+    .. note::
+
+        In more detail, the ``advanced_spectrum`` function also allows for
+        preprocessing of the QNode arguments before they are fed into the gates,
+        as long as this processing is *linear*. In particular, constant
+        prefactors for the encoding arguments are allowed.
+
+    **Example**
+
+    Consider the following example, which uses non-trainable inputs ``x``, ``y`` and ``z``
+    as well as trainable parameters ``w`` as arguments to the QNode.
+
+    .. code-block:: python
+
+        import pennylane as qml
+        import numpy as np
+        from advanced_spectrum import advanced_spectrum
```

Comment:
to match the API, I would either remove this and use `qml.fourier.advanced_spectrum` below, _or_ have `from pennylane.fourier import advanced_spectrum` here.